conf_file.cpp

```diff
@@ -16,268 +16,416 @@
  * limitations under the License.
  */
 
-#include "conf_file.h"
-
 #include <assert.h>
 #include <ctype.h>
-#include <stdio.h>
-#include <stdlib.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
 #include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
 
+#include "conf_file.h"
 #include "log.h"
 #include "util.h"
 
 struct config {
-    char *key;
-    char *value;
+    const char *key;
+    const char *value;
+    size_t key_len, value_len;
     struct config *next;
 };
 
 struct section {
-    char *name;
+    const char *name;
+    size_t len;
     struct config *configs;
-    struct config *current_config;
     struct section *next;
 };
 
+struct conffile {
+    int fd;
+    void *addr;
+    size_t len;
+    struct conffile *next;
+};
+
 ConfFile::~ConfFile()
 {
-    struct section *section;
-    struct config *config;
-
-    section = _sections;
-    while (section) {
-        struct section *tmp = section;
-
-        config = section->configs;
-        while (config) {
-            struct config *tmp_config = config;
-            free(config->key);
-            // config->value is on the same allocation as config->key, no need to free it
-            config = config->next;
-            free(tmp_config);
-        }
-
-        free(section->name);
-        section = section->next;
-        free(tmp);
-    }
+    release_all();
 }
 
-int ConfFile::parse_file()
+int ConfFile::parse(const char *filename)
 {
-    FILE *file;
-    char *entry = NULL;
-    int line = 0, ret = 0;
-    size_t n = 0;
-    ssize_t read;
+    int fd, ret = 0;
+    void *addr;
+    struct stat fstat;
+
+    assert(filename);
+
+    fd = open(filename, O_RDONLY | O_CLOEXEC);
+    if (fd < 0)
+        return -errno;
+
+    if (stat(filename, &fstat) < 0)
```

Comment:
Hint: this trick can be handy https://github.com/01org/mavlink-router/blob/master/logendpoint.cpp#L101-L102